### PR TITLE
Storage/list documents

### PR DESCRIPTION
**Pull request by YiMysty**
Please have a look at our [guidelines for contributions](https://github.com/Microsoft/AppCenter-SDK-Android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

A few sentences describing the overall goals of the pull request.

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 1]
```diff
@@ -258,9 +240,63 @@ public void onCallFailed(Exception e) {
...
+     * List (need optional signature to configure page size).
+     * The document type (T) must be JSON deserializable.
+     */
+    public <T> AppCenterFuture<PaginatedDocuments<T>> list(String partition) {
+        return getInstance().list(partition);
+    }
 
-    //region Replace implementation
+    /**
+     * Create a document
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: Dots missing

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 2]
```diff
@@ -258,9 +240,63 @@ public void onCallFailed(Exception e) {
...
+     */
+    private synchronized <T> AppCenterFuture<PaginatedDocuments<T>> instanceList(final String partition) {
+        final DefaultAppCenterFuture<PaginatedDocuments<T>> result = new DefaultAppCenterFuture<>();
+        getTokenAndCallCosmosDbApi(
+                partition,
+                result,
+                new TokenExchange.TokenExchangeServiceCallback() {
+
+                    @Override
+                    public void callCosmosDb(final TokenResult tokenResult) {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: final unnecessary here

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/client/CosmosDb.java [hunk 3]
```diff
@@ -94,23 +96,60 @@ public static String getDocumentBaseUrl(String databaseName, String collectionNa
                 DOCUMENT_DB_DOCUMENT_URL_PREFIX + (documentId == null ? "" : '/' + documentId);
     }
 
-    public static String GetDocumentUrl(TokenResult tokenResult, String documentId) {
+    public static String getDocumentUrl(TokenResult tokenResult, String documentId) {
         final String documentResourceIdPrefix = getDocumentBaseUrl(tokenResult.dbName(), tokenResult.dbCollectionName(), documentId);
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/client/CosmosDb.java]: final unnecessary here

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/client/CosmosDb.java [hunk 4]
```diff
@@ -94,23 +96,60 @@ public static String getDocumentBaseUrl(String databaseName, String collectionNa
...
         return getDocumentDbEndpoint(tokenResult.dbAccount(), documentResourceIdPrefix);
     }
 
+    public static synchronized <T> ServiceCall callCosmosDbListApi(
+            TokenResult tokenResult,
+            String continuationToken,
+            HttpClient httpClient,
+            ServiceCallback serviceCallback) {
+
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/client/CosmosDb.java]: add a comment or remove blank line

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Page.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.List;
+
+public class Page<T> {
+
+    /**
+     * Documents in the page.
+     */
+    private List<Document<T>> documents;
+
+    private DocumentError documentError;
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Page.java]: The getter is called `getError` and the backing field is `documentError`, rename field to error?

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Page.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    public Page() {
+
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Page.java]: remove blank line inside constructor

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        } else {
+           result.complete(new Page<T>(new NoSuchElementException()));
+        }
+        return result;
+    }
+
+    @Override
+    public Iterator<Document<T>> iterator() {
+        Iterator<Document<T>> it = new Iterator<Document<T>>() {
+            private int currentIndex = 0;
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java]: add a blank line

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public class PaginatedDocuments<T> implements Iterable<Document<T>> {
+
+    private Page<T> currentPage;
+
+    private TokenResult tokenResult;
+
+    private HttpClient httpClient;
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java]: Is this a gson document? might want to hide from serialization

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java [hunk 9]
```diff
@@ -9,13 +11,20 @@
...
 import com.microsoft.appcenter.utils.AppCenterLog;
 
+import java.util.ArrayList;
+import java.util.List;
+
 public final class Utils {
 
     public static final Gson sGson = new Gson();
 
+    public static final JsonParser parser = new JsonParser();
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: sParser, if its public, use private sField and expose getters instead

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java (hunk 8 above)
Comment (YiMysty) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java]: nope~ hmm do not know how to hide it....

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/client/CosmosDb.java (hunk 3 above)
Comment (YiMysty) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/client/CosmosDb.java]: did not see final key words :thinking:

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java (hunk 8 above)
Comment (guperrot) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java]: add transient

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Page.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.List;
+
+public class Page<T> {
+
+    /**
+     * Documents in the page.
+     */
+    @SerializedName(value = Constants.DOCUMENTS_FILED_NAME)
+    private List<Document<T>> documents;
+
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Page.java]: will add comment... seems extra line

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java (hunk 8 above)
Comment (YiMysty) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/PaginatedDocuments.java]: oki~

**Pull request merged**